### DISCUSSION TITLE
Current dir does not change when opening a file

### DISCUSSION THREAD

**Discussion by alextrastero** [nvim-tree/nvim-tree.lua#2062] [Q&A] [answered]
### Description

Steps:
`cd ~/Downloads`
`vim ../Documents/how-to-close-vim.docx`
`:NvimTreeOpen`

Result:
nvim tree opens in Downloads folder

Expected:
nvim tree opens in Documents folder

my config:
```
-- disable netrw at the very start of your init.lua (strongly advised)
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

nvim_tree.setup {
  respect_buf_cwd = true,
  sync_root_with_cwd = true,
}
```

### Neovim version

```text
NVIM v0.9.0-dev-1092+gbc15b075d
```

### Operating system and version

Linux fedora 6.1.14-200.fc37.x86_64

### nvim-tree version

Updated nvim-tree/nvim-tree.lua: bbb6d48..1d79a64

### Minimal config

```lua
-- disable netrw at the very start of your init.lua (strongly advised)
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

nvim_tree.setup {
  respect_buf_cwd = true,
  sync_root_with_cwd = true,
}
```

### Steps to reproduce

`cd ~/Downloads`
`vim ../Documents/how-to-close-vim.docx`
`:NvimTreeOpen`

### Expected behavior

nvim tree opens in Documents folder

### Actual behavior

nvim tree opens in Downloads folder

**Comment by Elias-Gill** [comment]
I was trying to solve this but is so annoying. 

As far as my limited knowledge about lua, nvimtree and programming itself allows me; the main problem is that the autocmd is running change_dir.fn() on every bufEnter, so when you focus nvimtree again, it switches to the first root directory of that instance.

It has to have a simple solution.

**Comment by alex-courtis** [comment]
`respect_buf_cwd` will cause nvim-tree to open at the root of the buffer/window's cwd. In this case, `~/Downloads`.

You can have it open the tree in `~/Documents` by first changing the local directory of the window / buffer via `:lcd ../Documents`.

`sync_root_with_cwd` will cause nvim-tree to change it's root to match the global cwd. In this case it is still `~/Downloads`

**Comment by alex-courtis** [comment]
I don't mean to be rude @alextrastero ; are you new to vim? The title of the document suggests so...

It appears we have a misunderstanding of expectations here.

In any case, you can handle startup directories etc. yourself: see [wiki: Open At Startup](https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup) for recipes to achieve this.

**Comment by CharlesARoy** [comment]
At risk of also being called "new to vim", I'm wondering if someone could spell out exactly how to achieve the behavior @alextrastero was trying to get. FWIW, when reading the documentation, I was also under the impression that respect_buf_cwd would behave as alextrastero described.

Anyway, I'm unsure of how to configure nvim-tree so that when you open it, it shows the contents of the directory of the file you currently have open in your buffer. 

Is there an autocommand that can be configured to do this? Something like: 
```
vim.api.nvim_create_autocmd({ "BufEnter" }, { callback = require("nvim-tree.api").tree.change_root(<command_to_load_cwd>)})
```

**Reply by alex-courtis** [reply]
> Anyway, I'm unsure of how to configure nvim-tree so that when you open it, it shows the contents of the directory of the file you currently have open in your buffer.

There are many ways to achieve this functionality, however this might be the clearest mapping:

```lua
vim.keymap.set("n", "<Leader>a", function()
  api.tree.open({
    path = vim.fn.fnamemodify(vim.api.nvim_buf_get_name(0), ":h"),
    find_file = true,
  })
end, { noremap = true })
```

It will open the tree at the directory of the current buffer and find it.

**Comment by alex-courtis** [comment]
> I'm wondering if someone could spell out exactly how to achieve the behavior @alextrastero was trying to get.

Very forcefully change the global cwd before opening, which will be in the new cwd:

```lua
vim.keymap.set("n", "<Leader>a", function()
  vim.cmd.cd(vim.fn.fnamemodify(vim.api.nvim_buf_get_name(0), ":h"))
  api.tree.open({
    find_file = true,
  })
end, { noremap = true })
```

Set `actions.change_dir.global` to have the global cwd changed when you change directories inside the tree e.g. `-` `<C-]>`

As above, there are many ways to achieve this behaviour. See https://github.com/nvim-tree/nvim-tree.lua/discussions/2350#discussioncomment-6647909

**Comment by alex-courtis** [comment]
Looking back over this discussion, it seems that the deficiency might be around `tree.open`, `tree.toggle`, `tree.find_file` behaviour for files that are within the current tree root but several levels down.

```help
*nvim-tree.update_focused_file.update_root*
(previously `update_focused_file.update_cwd`)
Update the root directory of the tree if the file is not under current
root directory. It prefers vim's cwd and `root_dirs`.
Otherwise it falls back to the folder containing the file.
Only relevant when `update_focused_file.enable` is `true`
  Type: `boolean`, Default: `false`
```

A means of always changing the tree root may be useful.

**Reply by alex-courtis** [reply]
#2405

**Reply by alex-courtis** [reply]
#2406

**Comment by CharlesARoy** [comment]
@alex-courtis thanks very much for your help with this, I appreciate it!

I was actually a little unclear in my question. What I really wanted was a single keymap that would both toggle nvim-tree and change to the directory of the file you currently have open in your buffer. That said, I was able to figure out the rest of it from your examples and came up with the following.

First, I created this function in my nvim-tree.lua file:
```lua
local M = {}
M.cwd_to_buf_and_toggle = function()
    local status_ok, nvimtree_api = pcall(require, "nvim-tree.api")
    if not status_ok then
        vim.notify("WARNING: nvim-tree.lua failed to load.")
        return
    end

    nvimtree_api.tree.find_file({ buf = vim.api.nvim_buf_get_name(0), update_root = true, focus = true })
    nvimtree_api.tree.toggle()
end
return M
```

Then, in my keymap file I have a keymap that calls the cwd_to_buf_and_toggle() function, e.g.:
```lua
vim.keymap.set("n", "<Leader>e", "<cmd>lua require('nvim-tree').cwd_to_buf_and_toggle()<CR>", silent=true)
```

Obviously, I don't need to export the function and define the keymap in a different file for this to work, but I like to have my keymaps all in one place for easier management.

**Reply by alex-courtis** [reply]
Nice one, glad it's working.

> but I like to have my keymaps all in one place for easier management.

Yes. I do the same.

**Comment by alexventuraio** [comment]
Today, I faced kind of the same need, being on my home directory I wanted to open the `nvim` config folder to work on, so I opened iTerm...

**...and from there, I did this:**
```
nvim ~/.config/nvim
:NvimTreeOpen
```

<img width="347" alt="Screenshot 2024-02-26 at 17 45 20" src="https://github.com/nvim-tree/nvim-tree.lua/assets/13169164/c35a4185-dc61-46e8-9df9-d44196e0e5ac">

**Result:**
Nvim-Tree opens in my Home folder 🤦🏻‍♂️

**Expected result:**
Nvim-Tree to have the `nvim` folder as the root one

---
Doing some googling, I found that there are a lot of people asking for the same thing and tried multiple suggestions but the one that really worked like a charm was [this one I found in Reddit](https://www.reddit.com/r/neovim/comments/o53wlw/starting_neovim_with_a_folder_argument_makes/) and that I tweaked a little bit for my Lua config:
```lua
vim.api.nvim_create_autocmd('VimEnter', {
  desc = 'Open Nvim-Tree on the given directory from command line',
  group = vim.api.nvim_create_augroup('NvimTreeSetGivenDirectory', { clear = true }),
  pattern = '*',
  command = "if argc() == 1 && isdirectory(argv()[0]) && !exists('s:std_in') | wincmd p | enew | execute 'cd '.argv()[0] | endif",
})
```

> P.D.:
> I'd love to convert the `command` into Lua code just because, so if anyone can help out, that'll be much appreciated (if possible of course). Thanks 😺

**Reply by alex-courtis** [reply]
Take a look at the [Open At Startup](https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup) recipes.

Lua code for your command looks to be all there.

**Comment by alex-courtis** [comment, marked as answer]
Please see [Open At Startup](https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup)
